### PR TITLE
For #3563. Use `ListAdapter` for Bookmark folder selection list.

### PR DESCRIPTION
**Pull request by dector**
- Using `ListAdapter` for bookmark folder selection.
  - Considering `node.guid` as item id.
  - Since only one folder can be selected, I'm searching for previously selected item in selection callback (operation complexity is __O(__ N __)__ but it's still better than `notifyDataSetChanged()`). Thus, changing only one or two items.

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] ~~**Tests**: This PR includes thorough tests or an explanation of why it does not~~
- [x] ~~**Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not~~
- [x] ~~**Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features~~

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/selectfolder/SelectBookmarkFolderAdapter.kt [hunk 1]
```diff
@@ -96,8 +98,36 @@ class SelectBookmarkFolderAdapter(private val sharedViewModel: BookmarksSharedVi
...
+    private inline fun BookmarkNode.isSelected(): Boolean =
+        this == sharedViewModel.selectedFolder
+
     companion object {
         private const val maxDepth = 10
         private const val dpsToIndent = 10
     }
 }
+
+private class DiffCallback : DiffUtil.ItemCallback<BookmarkNodeWithDepth>() {
```
Comment (NotWoods) [inline app/src/main/java/org/mozilla/fenix/library/bookmarks/selectfolder/SelectBookmarkFolderAdapter.kt]: nit: make this an `object`

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/selectfolder/SelectBookmarkFolderAdapter.kt [hunk 2]
```diff
@@ -96,8 +98,36 @@ class SelectBookmarkFolderAdapter(private val sharedViewModel: BookmarksSharedVi
...
+
+    override fun areItemsTheSame(
+        oldItem: BookmarkNodeWithDepth,
+        newItem: BookmarkNodeWithDepth
+    ) = (oldItem.node.guid == newItem.node.guid)
```
Comment (NotWoods) [inline app/src/main/java/org/mozilla/fenix/library/bookmarks/selectfolder/SelectBookmarkFolderAdapter.kt]: nit: remove parenthesis 
```suggestion
    ) = oldItem.node.guid == newItem.node.guid
```

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/selectfolder/SelectBookmarkFolderAdapter.kt [hunk 3]
```diff
@@ -16,15 +18,18 @@ import org.jetbrains.anko.image
...
-    private var tree: List<BookmarkNodeWithDepth> = listOf()
+    ListAdapter<BookmarkNodeWithDepth, BookmarkFolderViewHolder>(DiffCallback()) {
 
     fun updateData(tree: BookmarkNode?) {
-        this.tree = tree!!.convertToFolderDepthTree().drop(1)
-        notifyDataSetChanged()
+        val updatedData = tree
+            ?.convertToFolderDepthTree()
+            ?.drop(1)
+            ?: emptyList()
```
Comment (NotWoods) [inline app/src/main/java/org/mozilla/fenix/library/bookmarks/selectfolder/SelectBookmarkFolderAdapter.kt]: nit: use `.orEmpty()`

Comment (dector) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/selectfolder/SelectBookmarkFolderAdapter.kt]: Good suggestion, thanks!

**Pull request merged**